gobblin-yarn/src/main/java/gobblin/yarn/GobblinHelixJobLauncher.java

```diff
@@ -0,0 +1,273 @@
+ *
+ * @author ynli
+ */
+public class GobblinHelixJobLauncher extends AbstractJobLauncher {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(GobblinHelixJobLauncher.class);
+
+  private static final String WORK_UNIT_FILE_EXTENSION = ".wu";
+
+  private final HelixManager helixManager;
+  private final TaskDriver helixTaskDriver;
+  private final String helixQueueName;
+  private final String jobResourceName;
+
+  private final FileSystem fs;
+  private final Path appWorkDir;
+  private final Path inputWorkUnitDir;
+
+  private final int stateSerDeRunnerThreads;
+
+  public GobblinHelixJobLauncher(Properties jobProps, HelixManager helixManager, Path appWorkDir) throws Exception {
+    super(jobProps);
+
+    this.helixManager = helixManager;
+    this.helixTaskDriver = new TaskDriver(this.helixManager);
+
+    this.fs = FileSystem.get(new Configuration());
+    this.appWorkDir = appWorkDir;
+    this.inputWorkUnitDir = new Path(appWorkDir, ConfigurationConstants.INPUT_WORK_UNIT_DIR_NAME);
+
+    this.helixQueueName = this.jobContext.getJobName();
+    this.jobResourceName = TaskUtil.getNamespacedJobName(this.helixQueueName, this.jobContext.getJobId());
+
+    this.stateSerDeRunnerThreads = Integer.parseInt(jobProps.getProperty(ParallelRunner.PARALLEL_RUNNER_THREADS_KEY,
+        Integer.toString(ParallelRunner.DEFAULT_PARALLEL_RUNNER_THREADS)));
+  }
+
+  @Override
+  protected void runWorkUnits(List<WorkUnit> workUnits) throws Exception {
+    try {
+      submitJobToHelix(this.jobContext.getJobName(), this.jobContext.getJobId(),
+          createJob(this.jobContext.getJobName(), this.jobContext.getJobId(), workUnits));
+      waitForJobCompletion();
+      this.jobContext.getJobState().setState(JobState.RunningState.SUCCESSFUL);
+      this.jobContext.getJobState().addTaskStates(collectOutputTaskStates());
+    } finally {
+      deletePersistedWorkUnitsForJob(this.jobContext.getJobId());
+    }
+  }
+
+  @Override
+  protected JobLock getJobLock() throws IOException {
+    return new FileBasedJobLock(this.fs, this.jobProps.getProperty(ConfigurationKeys.JOB_LOCK_DIR_KEY),
+        this.jobContext.getJobName());
+  }
+
+  @Override
+  protected void executeCancellation() {
+    // Currently not supported yet
+  }
+
+  /**
+   * Create a job from a given batch of {@link WorkUnit}s.
+   */
+  private JobConfig.Builder createJob(String jobName, String jobId, List<WorkUnit> workUnits)
+      throws IOException {
+    Map<String, TaskConfig> taskConfigMap = Maps.newHashMap();
+
+    Closer closer = Closer.create();
+    try {
+      ParallelRunner stateSerDeRunner = closer.register(new ParallelRunner(this.stateSerDeRunnerThreads, this.fs));
+
+      for (WorkUnit workUnit : workUnits) {
+        if (workUnit instanceof MultiWorkUnit) {
+          // Flatten the MultiWorkUnit and add each individual WorkUnit
+          for (WorkUnit innerWorkUnit : JobLauncherUtils.flattenWorkUnits(((MultiWorkUnit) workUnit).getWorkUnits())) {
```

Comment:
I couldn't remember exactly why I did this initially but it had something to do with allowing Helix to have full control on the assignment/grouping of `WorkUnit`s. I think we can stick to this and see how Helix does in terms of load balancing.